ROADMAP.md

```diff
@@ -0,0 +1,157 @@
+
+Motivation: While analytic derivatives provide massive performance gains, 
+they also require significant user effort to implement. 
+This creates a high activation energy that prevents many users from taking advantage of the most powerful feature of OpenMDAO. 
+
+Overall Goal: Lower the effort required to implement analytic partial derivatives for components, 
+offering a spectrum of options that can potentially trade required user effort with computational efficiency. 
+
+## 1) Coloring applied to approximated partial derivatives 
+### Goal: 
+    - Usable for FD and CS approximations 
+    - Offers much higher performance for components with very sparse partial derivative Jacobians 
+      (e.g. vectorized components with no interaction between vector elements)
+    - When used with CS, effectively offers a fast-forward mode AD
+
+### Potential Challenges: 
+    - Computational cost of coloring may be excessive for models with a lot of of instances 
+    - Coloring algorithms are not perfect, and may result in an invalid Jacobian, 
+      so we need an effective way for users to check their colored partials. 
+      (the existing check_partials functionality can be leveraged, but may need some updates)
+    - In some cases, sparsity may not be high enough to offer meaningful performance gains. 
+      How can a user check this? 
+
+### Notes: 
+    - Prototype implementation already available in experimental features 
+    - Partial coloring algorithm will also be useful for AD partials
+
+
+## 2) Algorithmic differentiation for component partial derivatives
+    
+### Goal: 
+- Forward and reverse mode AD that works for a wide variety of general use cases including many numpy functions 
+- Relatively good performance compared to hand differentiation
+
+### Potential Challenges: 
+- AD tools will struggle with current OpenMDAO syntax for `compute` and `apply_nonlinear` methods, 
+  so some kind of translation layer will be needed
+- Python AD libraries are not as well developed as other languages (e.g. C, Fortran, Julia)
+- Certain coding practices are not compatible with AD, and will need to be avoided (e.g. modification of instance attributes, functions with side-effects)
+
+
+----------------------------------------------
+----------------------------------------------
+----------------------------------------------
+# Establishing A Community Contribution Process
+
+## 1) Planning for the 2020 OpenMDAO workshop 
+
+## 2) POEMs: **p**roposal for **O**penMDAO **e**nhance**m**ent
+- Loosely based on the model used by cPython project (PEP process)
+- A new repo has been created for tracking POEMs: 
+  http://github.com/openmdao/POEMs
+- Notes: We are planning to work the POEM process on PR 1086 (https://github.com/OpenMDAO/OpenMDAO/pull/1086)
+
+## 3) Establishing a formal OpenMDAO plugin system
+- To go alone with the POEM process, 
+we need a method for users to add functionality without merging their code to core codebase. 
```

Comment:
Hooray extensions! I learned a lot from watching [this talk](https://www.youtube.com/watch?v=7K72DPDOhWo).

Since then, we've built a couple such systems: getting it wrong is worse than not having it, as forcing people to contribute to the core, where everything can be tested and released together, has a tyrannical elegance to it. But if plugins are The Future, having them be robustly verifiable, testable, documented, and dogfooded can work.

Some ideas (maybe for here, maybe as POEM, which I can't guarantee being able to shepherd):
- for each thing to make pluggable (it doesn't have to be one big PR)
  - introduce a versioned [`entry_point`](https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins) which an extension author can implement in `setup.(py|cfg)`, e.g.
    ```ini
    [options.entry_points]
    openmdao_solver_v0 = 
        foo_package_foo_solver = foo_package.solver:FooSolver
        # or
        foo_package_solver_foo = foo_package.solver:get_solver
    ```
    - there are other approaches, but this one's built-in
  - refactor all existing implementations of the current pluggable to also use those `entry_point`s e.g. 
    ```ini
    [options.entry_points]
    openmdao_solver_v0 = 
        openmdao_solver_scipy_krylov = openmdao.solvers.linear.scipy_iter_solver.ScipyKrylov
        openmdao_solver_newton = openmdao.solvers.nonlinear.newton.NewtonSolver       
        ... 
    ```
    - this is important, if only for there to be non-trivial examples
  - hoist the base that `FooSolver` would implement to `api.py` with implementation-free, heavily documented, versioned [`ABC`](https://docs.python.org/3/library/abc.html#abc.ABC)s, e.g. so `Solver(SolverABC0)`
    - better than "just" more docs are [PEP 484](https://www.python.org/dev/peps/pep-0484/) annotations
      - ...but only if tested in CI
  - add a (tested) tutorial documentation for each plugin type
  - at runtime, load them all with [`pkg_resources.iter_entry_points`](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#convenience-api)
    - this is a little slow
    - using [`entrypoints`](https://github.com/takluyver/entrypoints) is slightly faster
  - consider getting a versioned [trove classifier](https://github.com/pypa/warehouse/issues/new?title=Request+trove+classifier&template=new-trove-classifier.md) for pypi, and encourage developers and end users to use it